### DISCUSSION TITLE
is there a mismatch between urqls `Operation` and  graphql-ws `SubscribePayload` ?

### DISCUSSION THREAD

**Discussion by 97amarnathk** [enisdenjo/graphql-ws#166] [Q&A] [answered]
While using graphql-ws with urql , I notice that there is a type mismatch between the SubscribePayload  received by graphql-ws server and what urql is sending to it.

According to graphql-ws, the type is 

```typescript
export interface SubscribePayload {
    readonly operationName?: string | null;
    readonly query: string;
    readonly variables?: Record<string, unknown> | null;
}
```

And according to urql it is 

```typescript
export interface Operation<Data = any, Variables = any>
  extends GraphQLRequest<Data, Variables> {
  readonly kind: OperationType;
  context: OperationContext;
}
```

And this operation is being passed directly to the graphql-ws client as:

```typescript
forwardSubscription: (operation) => {
  return {
     subscribe: (sink) => {
     const dispose = subscriptionClient.subscribe(operation, sink);
     return {
      unsubscribe: dispose,
    };
   },
};
```

**Comment by enisdenjo** [comment, marked as answer]
The `Operation` interface extends [`GraphQLRequest`](https://github.com/FormidableLabs/urql/blob/b6890ba1b0c9fa9e99301338555ea10d602cee22/packages/core/src/types.ts#L27) which has the necessary fields for the payload (`query` and `variables`).

**Reply by 97amarnathk** [reply]
Oh. Okay. So this means that `context` from urql Operation is not supposed to be passed to the client.

So how do I pass the JWT access token on each graphql subscription (websocket is already connected)?

In the examples, I see that in the server side , we can check the `request.headers.authToken`  inside `onSubscribe`.
How do I set this `authToken` in the request header for the web socket in the client
? I find that Websockets do not allow you to define custom headers.

**Reply by enisdenjo** [reply]
The `context`, and all other extraneous fields, are passed, but are completely ignored by the client and the server.

Authentication details are sent along the [`ConnectionInit` message](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#connectioninit). You use [the client `connectionParams` option](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.clientoptions.md#connectionparams) for setting them, and [the server `context.connectionParams`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.context.md#connectionparams) for reading them.

You say that the examples check for the `request.headers.authToken`, may I ask where exactly?

**Reply by 97amarnathk** [reply]
I want to authenticate each time the user makes a Graphql operation .
So my understanding is that I should authenticate in `onSubscribe`.

To do this, I need to pass the authToken via websocket for each client request on the connected websocket.

I see this also being done in the example: https://github.com/enisdenjo/graphql-ws#ws-auth-handling

Check the onSubscribe method, it is calling `handleAuth`, which checks for authToken in request headers.

My doubt is **how do I add this authToken in the client end**

**Reply by enisdenjo** [reply]
Ah yes, that is just an example of how to implement a custom handler. You can indeed set HTTP WS upgrade headers, but sadly not in the browser.

As said before, the authentication parameters are passed through the `connectionParams` on the client and the same are set in the `context.connectionParams` on the server. This `context` value is passed as the first argument for every callback in the server, you're free to validate the `connectionParams` whenever you want. Please read the documentation.

**Reply by n1ru4l** [reply]
This works for me:

On the client pass in additional arg as a subscribe object property.

```ts
wsClient.subscribe({ ...graphQLParams, authToken: "test" }, sink)
````
Server with `onSubscribe`

```ts
    onSubscribe: async (_, msg) => {
      console.log(msg);
      const session = await authenticate(msg.authToken)
      const args: ExecutionArgs = {
        schema,
        operationName: msg.payload.operationName,
        document:
          typeof msg.payload.query === "object"
            ? msg.payload.query
            : parse(msg.payload.query),
        variableValues: msg.payload.variables,
        contextValue: {...context, session },
      };

      // don't forget to validate when returning custom execution args!
      const errors = validate(args.schema, args.document, validationRules);
      if (errors.length > 0) {
        return errors; // return `GraphQLError[]` to send `ErrorMessage` and stop subscription
      }

      return args;
    },
```

**Reply by 97amarnathk** [reply]
Yes, but in my case,
In my case it is important to authenticate each websocket request individually. since accessToken can change over time,even after ConnectionInit has happened.

**Reply by n1ru4l** [reply]
Ideally, SubscribePayload could be typed for the additional props as:

```
type SubscribePayload<TExtensions = {}> = {
    readonly operationName?: string | null;
    readonly query: string;
    readonly variables?: Record<string, unknown> | null;
} & TExtensions

**Reply by n1ru4l** [reply]
> Yes, but in my case,
> In my case it is important to authenticate each WebSocket request individually. since accessToken can change over time, even after ConnectionInit has happened.

That is exactly what my snippet does?

```
     * The subscribe callback executed right after
     * acknowledging the request before any payload
     * processing has been performed.
```

onSubscribe is invoked for every operation

**Reply by 97amarnathk** [reply]
SubscribePayload with additional props seems like a good idea to me as well.

I would be happy to raise a PR for this over the weekend.

**Reply by n1ru4l** [reply]
Yeah it already works only the compiler will complain, something like this might work:

```ts
type SubscribePayload<TExtensions = Record<string, unknown>> = {
  readonly operationName?: string | null;
  readonly query: string;
  readonly variables?: Record<string, unknown> | null;
} & TExtensions
```

**Reply by enisdenjo** [reply]
The way I would recommend solving the token changing over time is simply kicking the client off once it becomes invalid. The client is built to silently reconnect, and the client connectionParams is a function called on every connect. You just provide the latest token for each time connectionParams gets invoked.

What you do is validate the context connectionParams for every onSubscribe (or whenever you want), if it has expired, you simply `context.extra.socket.close(4401, 'Unauthorized')`. After the connection has closed, the client will silently retry invoking the `connectionParams` function, picking up the new token and reconnecting successfully on the server. All subscriptions will resubscribe automatically - the client will quite literally notice nothing.

This way you keep the subscriptions messages lean (you dont add the long token on every subscribe) and you're very secure because the whole connection is re-evaluated. Also, what you're doing, if you continue with passing all possible authentication details on every message, is simulating HTTP - you might just as well use Server-Side Events...

Furthermore, you must be aware that WebSocket connections cannot be hijacked. You can be 100% certain that the client stays the same during the whole connection life.

**Reply by enisdenjo** [reply]
Hey guys, consider checking the new recipe about [ws server and client auth usage with token expiration, validation and refresh](https://github.com/enisdenjo/graphql-ws#auth-token). It showcases my explanation above in code.

**Reply by 97amarnathk** [reply]
A minor problem is that 4401 is also a fatal exception by default. 
https://github.com/enisdenjo/graphql-ws/blob/master/src/client.ts#L472-L479

I am using 4430 (or any other code reserved for application and not used by graphql-ws)
https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent

**Reply by enisdenjo** [reply]
My bad! I changed the example to [use `4403: Forbidden` instead](https://github.com/enisdenjo/graphql-ws/commit/7642666be4e1e1e6d6394fbadb779f83a461cd7c).

Furthermore, I recommend using `4403` because [returning false from `onConnect` automatically closes the connection with a `4403: Forbidden`](https://github.com/enisdenjo/graphql-ws/blob/7642666be4e1e1e6d6394fbadb779f83a461cd7c/src/server.ts#L535-L537).

**Reply by jyshin97** [reply]
@enisdenjo thank you so much for the lib! I noticed in the [example](https://github.com/enisdenjo/graphql-ws#auth-token) you still had

 ```tokenExpiryTimeout = setTimeout(() => {
        if (socket.readyState === WebSocket.OPEN)
          socket.close(CloseCode.Unauthorized, 'Unauthorized');
      }, getCurrentTokenExpiresIn());
```

when I'm assuming it should be 

 ```tokenExpiryTimeout = setTimeout(() => {
        if (socket.readyState === WebSocket.OPEN)
          socket.close(CloseCode.Forbidden, 'Forbidden');
      }, getCurrentTokenExpiresIn());
```

**Reply by enisdenjo** [reply]
Thanks @jyshin97 for pointing that out, fixed in https://github.com/enisdenjo/graphql-ws/commit/1aee1e87ec30361c3e174607f60e1bb53751be69. 😄
